File: trees/0112-path-sum.go
package trees

import "github.com/shahzodshafizod/leetcode/pkg"

// https://leetcode.com/problems/path-sum/

func hasPathSum(root *pkg.TreeNode, targetSum int) bool {
	if root == nil {
		return false
	}
	targetSum -= root.Val
	isLeaf := true
	for _, child := range []*pkg.TreeNode{root.Left, root.Right} {
		if child == nil {
			continue
		}
		isLeaf = false
		if hasPathSum(child, targetSum) {
			return true
		}
	}
	return targetSum == 0 && isLeaf
}

// func hasPathSum(root *pkg.TreeNode, targetSum int) bool {
// 	var dfs func(node *pkg.TreeNode, currSum int) bool
// 	dfs = func(node *pkg.TreeNode, currSum int) bool {
// 		if node == nil {
// 			return false
// 		}
// 		currSum += node.Val
// 		if node.Left == nil && node.Right == nil {
// 			return currSum == targetSum
// 		}
// 		return dfs(node.Left, currSum) || dfs(node.Right, currSum)
// 	}
// 	return dfs(root, 0)
// }

File: trees/0112-path-sum.py
import unittest
from typing import Optional
from design.tree import TreeNode, create_tree

# https://leetcode.com/problems/path-sum/

# python3 -m unittest trees/0112-path-sum.py

class Solution(unittest.TestCase):
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        def dfs(node: TreeNode, currSum: int) -> bool:
            if not node:
                return False
            currSum += node.val
            if not node.left and not node.right:
                return currSum == targetSum
            return dfs(node.left, currSum) or dfs(node.right, currSum)
        return dfs(root, 0)

    def test(self) -> None:
        for root, targetSum, expected in [
            ([5,4,8,11,None,13,4,7,2,None,None,None,1], 22, True),
            ([1,2,3], 5, False),
            ([], 0, False),
            ([1,2], 1, False),
            ([1,2,None,3,None,4,None,5], 6, False),
            ([1,2,None,3,None,4,None,5], 15, True),
        ]:
            root = create_tree(root)
            output = self.hasPathSum(root, targetSum)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")
